python/lsst/daf/butler/gen2convert/writer.py

```diff
@@ -0,0 +1,316 @@
+                    log.debug("No StorageClass name configured for %s; trying unqualified python type '%s'",
+                              gen2dst.name, unqualified)
+                    try:
+                        storageClass = scFactory.getStorageClass(unqualified)
+                    except KeyError:
+                        log.warn("No StorageClass found for %s; skipping.", gen2dst.name)
+                        continue
+            log.debug("Using StorageClass %s for %s", storageClass.name, gen2dst.name)
+            self.datasetTypes[datasetTypeName] = DatasetType(
+                name=datasetTypeName,
+                storageClass=storageClass,
+                dataUnits=translators[datasetTypeName].gen3units
+            )
+        converted = ConvertedRepo(gen2repo, camera=camera, run=run, translators=translators)
+        # Add parent repositories first, so self.repos is sorted topologically.
+        for parent in gen2repo.parents:
+            self._addConvertedRepoSorted(parent)
+        # Now we can finally add the current repo to self.repos.
+        self.repos[gen2repo.root] = converted
+        return converted
+
+    def run(self, registry, datastore):
+        """Main driver for ConversionWriter.
+
+        Runs all steps to create a Gen3 Repo, aside from Camera registration
+        (we merely check that the needed Cameras, Sensors, and PhysicalFilters
+        have already been registered).
+        """
+        self.checkCameras(registry)
+        self.insertSkyMaps(registry)
+        self.insertObservations(registry)
+        self.insertDatasetTypes(registry)
+        self.insertDatasets(registry, datastore)
+        # TODO: associate parent repo datasets with child repo Collections
+
+    def checkCameras(self, registry):
+        """Check that all necessary Cameras are already present in the
+        Registry.
+        """
+        log = Log.getLogger("lsst.daf.butler.gen2convert")
+        cameras = set()
+        for repo in self.repos.values():
+            cameras.add(self.config["mappers"][repo.gen2.MapperClass.__name__]["camera"])
+        for camera in cameras:
+            log.debug("Looking for preexisting Camera '%s'.", camera)
+            if registry.findDataUnitEntry("Camera", {"camera": camera}) is None:
+                raise LookupError(
+                    "Camera '{}' has not been registered with the given Registry.".format(camera)
+                )
+
+    def insertSkyMaps(self, registry):
+        """Add all necessary SkyMap DataUnits (and associated Tracts and
+        Patches) to the Registry.
+        """
+        log = Log.getLogger("lsst.daf.butler.gen2convert")
+        for sha1, skyMap in self.skyMaps.items():
+            skyMapName = self.skyMapNames.get(sha1, None)
+            try:
+                existing, = registry.query("SELECT skymap FROM SkyMap WHERE sha1=:sha1",
+                                           sha1=sha1)
```

Comment:
In general we probably don't want to rely on `query` too much, when other API calls should exist.